stake-pool/program/src/processor.rs

```diff
@@ -892,20 +943,75 @@ mod tests {
             withdraw_authority_key,
             fee,
             owner_key,
-            owner_fee_key,
-            owner_fee_account,
+            owner_account,
+            owner_fee_key: token.key,
+            owner_fee_account: token.account,
             mint_key,
             mint_account,
         }
     }
 
+    fn do_deposit(pool_info: &mut StakePoolInfo, stake_balance: u64) -> DepositInfo {
+        // Create account to receive minted tokens
+        let mut token = create_token_account(
+            &TOKEN_PROGRAM_ID,
+            &pool_info.mint_key,
+            &mut pool_info.mint_account,
+        );
+        do_deposit_custom(pool_info, stake_balance, &mut token)
+    }
+
+    fn do_deposit_custom(
+        pool_info: &mut StakePoolInfo,
+        stake_balance: u64,
+        token: &mut TokenInfo,
+    ) -> DepositInfo {
+        let stake_account_key = Pubkey::new_unique();
+        let mut stake_account_account = Account::new(stake_balance, 100, &stake_program_id());
+        // TODO: Set stake account Withdrawer authority to pool_info.deposit_authority_key
+
+        // Call deposit
+        let _result = do_process_instruction(
+            deposit(
+                &STAKE_POOL_PROGRAM_ID,
+                &pool_info.pool_key,
+                &pool_info.deposit_authority_key,
+                &pool_info.withdraw_authority_key,
+                &stake_account_key,
+                &token.key,
+                &pool_info.owner_fee_key,
+                &pool_info.mint_key,
+                &TOKEN_PROGRAM_ID,
+                &stake_program_id(),
+            )
+            .unwrap(),
+            vec![
+                &mut pool_info.pool_account,
+                &mut Account::default(),
+                &mut Account::default(),
+                &mut stake_account_account,
+                &mut token.account,
+                &mut pool_info.owner_fee_account,
+                &mut pool_info.mint_account,
+                &mut Account::default(),
+                &mut Account::default(),
+                &mut Account::default(),
+            ],
+        )
+        .expect("Error on stake pool deposit");
+
+        DepositInfo {
+            stake_account_key,
+            stake_account_account,
+            token_receiver_key: token.key,
+            token_receiver_account: token.account.clone(),
+            token_owner_key: token.owner,
```

Comment:
Actually there are two deposit functions: one accepts token account as a parameter and another creates new token account and passes it to the first function (`do_deposit` and `do_deposit_custom`). We can make `do_deposit_custom` return only stake account info (key/account) but it is better to have them both return the same DepositInfo struct (to make code calling these two functions similar).

Reply:
I see what you mean.  To be honest, it was a bit hard to follow some of the tests because of the account creation in `do_deposit`, so I would explicitly create the account outside, and only have `do_deposit_custom`.  What do you think?

Reply:
Actually it was the other way round :) We did do_deposit, which is used in deposit/withdraw/claim tests. And then realized that we need to collect tokens for two deposits to the same accounts, so we've added do_deposit_custom to allow that. And I think in remaining tests we will mostly use do_deposit which will create token account by itself.

But you of course have the last word, if you think it will be clearer if token receiver was created outside of the deposit we will do it this way :)

Reply:
Creating an account in `do_deposit` still seems confusing to me, so I'd prefer that we only have the version that takes `&mut TokenInfo`, and then also removing the `token_*` fields from `DepositInfo`.  That way, you don't need to worry about data duplication.